generator/sequence_transform_data.py

```diff
@@ -495,40 +531,64 @@ def traverse(trie_node):
     # quiet_print(f'{err(0)} Data "{cyan(table)}"')
 
     def serialize(node: Dict[str, Any]) -> List[int]:
-        data = node['data']
+        # print(f"{json.dumps(node, indent=4)}")
+        data = []
+        # data = node['data']
         # quiet_print(f'{err(0)} Serialize Data "{cyan(data)}"')
+        if 'node_header_data' in node:
+            data = data + node['node_header_data']
+
+        if 'match_data' in node:
+            data = data + node['match_data']
 
-        if not node['links']:  # Handle a leaf table entry.
-            return data
+        if 'chain_data' in node:
+            for cmatch, _ in node['chain_data']:
+                data = data + encode_link(cmatch['SUB_RULE']) + \
+                    cmatch['DATA']
 
-        elif len(node['links']) == 1:  # Handle a chain table entry.
-            return data + [1] + [symbol_map[c] for c in node['chars']] + [0]
+        if 'str' in node:  # Handle a chain table entry.
+            return data + [1] + [symbol_map[c] for c in node['str']] + [0]
 
-        else:  # Handle a branch table entry.
+        if 'chars' in node:  # Handle a branch table entry.
             links = [TRIE_BRANCH_BIT]
 
             for c, link in zip(node['chars'], node['links']):
-                links += [symbol_map[c]] + encode_link(link)
+                links += [symbol_map[c]] + encode_link(link['node'])
 
             return data + links + [0]
 
+        return data
+
     uint16_offset = 0
 
     # To encode links, first compute byte offset of each entry.
     for table_entry in table:
-        table_entry['uint16_offset'] = uint16_offset
+        table_entry['node']['OFFSET'] = uint16_offset
+        temp_uint16_offset = uint16_offset + len(table_entry.get('node_header_data', []))
+        if 'match_data' in table_entry:
+            table_entry['match_node']['OFFSET'] = temp_uint16_offset
+            temp_uint16_offset += len(table_entry['match_data'])
+        if 'chain_data' in table_entry:
+            # print(f"offset chain_data {table_entry['chain_data']}")
+            for cmatch, cnode in table_entry['chain_data']:
+                cnode['OFFSET'] = temp_uint16_offset + 2
+                temp_uint16_offset += 2 + len(cmatch['DATA'])
+
         uint16_offset += len(serialize(table_entry))
 
         assert 0 <= uint16_offset <= 0xffff
 
     # Serialize final table.
-    return [b for node in table for b in serialize(node)]
+    trie_data = [b for node in table for b in serialize(node)]
+
+    return trie_data
```

Comment:
any reason to do that?

Reply:
Yes, I there assignment to `trie_data` changes the offsets of the items in `table`. Since I wanted to print out `table` after those changes, I needed a variable so that I could sneak in the print before returning.

I left it there because it's nice to have a name for what we are returning, since the expression is non trivial. Also, I might read that debug print after making a smart print function that can prune the output down to only the parts we care about.